### PR TITLE
Add hint for similar function parameter names when they don't match

### PR DESCRIPTION
**Pull request by laurenceisla**
Related to #2571 and #2451.

### CODE REVIEW

File: src/PostgREST/Error.hs [hunk 1]
```diff
@@ -17,6 +17,8 @@ module PostgREST.Error
 
 import qualified Data.Aeson                as JSON
 import qualified Data.ByteString.Char8     as BS
+import qualified Data.FuzzySet             as F
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:20]: To make the code more clear
```suggestion
import qualified Data.FuzzySet             as Fuzzy
```

File: src/PostgREST/Error.hs [hunk 2]
```diff
@@ -183,13 +186,24 @@ instance JSON.ToJSON ApiRequestError where
...
+-- TODO: adapt for a Content-Type other than application/json or limit only to that header
+noRpcHint :: Text -> Text -> Text -> ProcsMap -> [ProcDescription] -> Text
+noRpcHint schema procName params allProcs overloadedProcs =
+  maybe ("No functions with a similar name found in the schema " <> schema) -- Should be null and adapt to function found but no similar parameters found
+        (("Perhaps you meant to call the function " <> schema <> ".") <>)
+        possibleProcs
+  where
+    possibleProcs
+      | null overloadedProcs = F.getOne (F.fromList [qiName k | k <- HM.keys allProcs, qiSchema k == schema]) procName
+      | otherwise            = (procName <>) <$> F.getOne (F.fromList ["(" <> T.intercalate ", " [ppName prm | prm <- pdParams ov] <> ")" | ov <- overloadedProcs]) params
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:205]: Some variables would make this clearer

File: src/PostgREST/Error.hs [hunk 3]
```diff
@@ -37,12 +39,14 @@ import qualified PostgREST.MediaType        as MediaType
...
 import PostgREST.SchemaCache.Proc         (ProcDescription (..),
-                                           ProcParam (..))
+                                           ProcParam (..), ProcsMap)
 import PostgREST.SchemaCache.Relationship (Cardinality (..),
                                            Junction (..),
                                            Relationship (..))
 import Protolude
 
+-- >>>
+-- >>>
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:49]: Hm, why is this needed? Can it be deleted?

Comment (laurenceisla) [reply src/PostgREST/Error.hs:49]: Oops, forgot to remove those.

File: src/PostgREST/Error.hs [hunk 4]
```diff
@@ -170,26 +174,78 @@ instance JSON.ToJSON ApiRequestError where
...
+-- in the same schema and shows the best match as hint.
+--
+-- >>> :set -Wno-missing-fields
+-- >>> let procsMap = HM.fromList [(QualifiedIdentifier "api" "test" , []), (QualifiedIdentifier "api" "another" , []), (QualifiedIdentifier "private" "other" , [])]
+--
+-- >>> noRpcHint "api" "testt" "(val, param, name)" procsMap []
+-- Just "Perhaps you meant to call the function api.test"
+--
+-- >>> noRpcHint "api" "other" "()" procsMap []
+-- Just "Perhaps you meant to call the function api.another"
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:213]: Ah, this one doesn't work because `other` is on a non-exposed schema.

File: src/PostgREST/Error.hs [hunk 5]
```diff
@@ -170,26 +174,78 @@ instance JSON.ToJSON ApiRequestError where
...
+--
+-- >>> noRpcHint "api" "noclosealternative" "()" procsMap []
+-- Nothing
+--
+-- If a function is found with the given name, but no params match, then it does a fuzzy searh
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:218]:
```suggestion
-- If a function is found with the given name, but no params match, then it does a fuzzy search
```

File: src/PostgREST/Error.hs [hunk 6]
```diff
@@ -170,26 +174,78 @@ instance JSON.ToJSON ApiRequestError where
...
+-- to all the overloaded functions' params using the form "(param1, param2, param3, ...)"
+-- and shows the best match as hint.
+--
+-- >>> let procsDesc = [ProcDescription {pdParams = [ProcParam {ppName="val"}, ProcParam {ppName="param"}, ProcParam {ppName="name"}]}, ProcDescription {pdParams = [ProcParam {ppName="id"}, ProcParam {ppName="attr"}]}]
+--
+-- >>> noRpcHint "api" "test" "(vall, pqaram, nam)" procsMap procsDesc
+-- Just "Perhaps you meant to call the function api.test(val, param, name)"
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:225]: This is really cool!

Q: Why do the params use a Text type`"(vall, pqaram, nam)"`? Why can't they be a List?

Comment (laurenceisla) [reply src/PostgREST/Error.hs:225]: The text type is needed because fuzzy search looks for a Text in a set of possible matches, that's why the solution I saw was formatting the parameters as Text. So, behind the scenes it does this:

`Fuzzy.getOne (Fuzzy.fromList ["(val, param, name)", "(id, attr)"])  "(vall, pqaram, nam)"`

And returns the first one directly to be used by the hint. Although, the arguments could be formatted into a Text inside the function itself, but I wanted to reuse the `prms` function that already did that.

Comment (steve-chavez) [reply src/PostgREST/Error.hs:223]: Hm, I still don't understand. From the [fuzzy set docs](https://hackage.haskell.org/package/fuzzyset-0.2.3/docs/Data-FuzzySet.html) I don't see that they use that format.
 
```haskell
Fuzzy.getOne (Fuzzy.fromList ["(val, param, name)", "(id, attr)"]) "(vall, pqaram, nam)"
```

So the fuzzy search is ignoring the parentheses?

File: src/PostgREST/Error.hs (hunk 4 above)
Comment (laurenceisla) [reply src/PostgREST/Error.hs]: Yes.

File: src/PostgREST/Error.hs (hunk 6 above)
Comment (laurenceisla) [reply src/PostgREST/Error.hs]: Yes, it's doing that. I separated that login in other function and added a comment to hopefully make it clearer.

File: src/PostgREST/Error.hs [hunk 7]
```diff
@@ -170,26 +171,82 @@ instance JSON.ToJSON ApiRequestError where
...
-        (_, True, MTApplicationJSON) -> prms <> " function or the " <> schema <> "." <> procName <>" function with a single unnamed json or jsonb parameter"
-        _                            -> prms <> " function") <>
+        (_, True, MTApplicationJSON) -> fmtParams <> " or the " <> func <>" function with a single unnamed json or jsonb parameter"
+        _                            -> fmtParams) <>
       " in the schema cache"),
     "details" .= JSON.Null,
-    "hint"    .= ("If a new function was created in the database with this name and parameters, try reloading the schema cache." :: Text)]
+    "hint"    .= if hasPreferSingleObject || (isInvPost && contentType `elem` [MTTextPlain, MTTextXML, MTOctetStream])
+                 then Nothing
+                 else noRpcHint schema procName argumentKeys allProcs overloadedProcs ]
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:192]: Here the assumption is that if we're in the case of the single unnamed params we don't add the `hint` right? This should be clarified with a comment.

**Review by steve-chavez — APPROVED** [review]
